File: src/main/java/org/vanautrui/languages/compiler/parsing/astnodes/nonterminal/upperscopes/AST_Whole_Program.java
package org.vanautrui.languages.compiler.parsing.astnodes.nonterminal.upperscopes;

import org.vanautrui.languages.compiler.parsing.IASTNode;

import java.util.HashSet;
import java.util.Set;
import java.util.stream.Collectors;

public final class AST_Whole_Program implements IASTNode {

	//this contains all namespace nodes for the whole program

	public Set<NamespaceNode> namespaceNodeList = new HashSet<>();

	public AST_Whole_Program() { }

	public AST_Whole_Program(final NamespaceNode myNamespace) {
		//utility method, to facilitate creating AST_Whole_Program from a single namespace node
		this.namespaceNodeList.add(myNamespace);
	}

	@Override
	public String toSourceCode() {
		return this.namespaceNodeList
				.stream()
				.map(node -> node.toSourceCode())
				.collect(Collectors.joining("\n"));
	}
}

File: dragon-parser/src/main/compiler/parsing/astnodes/nonterminal/upperscopes/AST_Whole_Program.cpp

//standard headers
#include <vector>
#include <set>

//project headers
#include "AST_Whole_Program.hpp"

public final class AST_Whole_Program implements IASTNode {

	//this contains all namespace nodes for the whole program

	public final Set<NamespaceNode> namespaceNodes = new HashSet<>();

	public AST_Whole_Program() {
	}

	public AST_Whole_Program(final NamespaceNode myNamespace) {
		//utility method, to facilitate creating AST_Whole_Program from a single namespace node
		this.namespaceNodes.add(myNamespace);
	}

	public AST_Whole_Program(final TokenList tokens, final String namespace, boolean debug) throws Exception {

		//use this constructor as a shortcut to create an AST
		//if there is just a single namespace

		if (debug) {
			System.out.println("parse AST_Whole_Program");
		}

		this.namespaceNodes.add(new NamespaceNode(tokens, namespace, debug));
	}

}
